### PR TITLE
UI freeze mitigations for GTK implementation

### PR DESCRIPTION
**Pull request by weisJ**
Relates to #57 

I suspect that the GTK application thread wasn't dying got stuck waiting to be woken up. @zliuva could you take a look at this?

Also per the GTK specification calling `run` on `GTK::Application` after `quit` is unspecified behaviour, so instead of caching the instance a new one is created each time.

### CODE REVIEW

**Comment by zliuva** [comment]
@weisJ your description makes a lot of sense to me. I guess I have never not had the plugin running 24/7 :)

**Comment by weisJ** [comment]
Based on this https://github.com/weisJ/auto-dark-mode/issues/57#issuecomment-1150311779 the issue seems to be happening reliably. I guess it doesn't reproduce for you? Do you have an idea why `Gtk::Application::quit` doesn't immedeatly stop the execution of  `Gtk::Application::run`? My hunch is that it somehow doesn't wake up and then eventually will, due to some event happening. Hence the deviation in freeze time. How would I go about forcing the application to wake up and process signals?

**Comment by zliuva** [comment]
let me see if I can reproduce. 

what I meant was that I tend to keep IntelliJ open 24/7 so did not get to execercise this codepath in my daily use. now knowing it is triggered by disabling/exiting I can try reproduce.

File: linux/gtk/src/main/cpp/DarkModeGtk.cpp [hunk 1]
```diff
@@ -63,21 +65,24 @@ struct EventHandler {
...
-        app->hold();
         app->run();
     }
 
     EventHandler(JavaVM *jvm_, jobject callback_) {
         jvm = jvm_;
         callback = callback_;
 
+        app = Gtk::Application::create();
+        app->hold();
```
Comment (zliuva) [inline linux/gtk/src/main/cpp/DarkModeGtk.cpp:84]: we should also move `settings = Gtk::Settings::get_default();` to be after `app` has been created. (otherwise you get null it seems)

**Review by zliuva — COMMENTED** [review]
I think moving `app` from init to the eventhandler itself is a good move (for cases such as disabling then renabling the plugin. I'm not sure if `init` is called again in that case).


HOWEVER, I still experience the same issue in #57 ( on my setup the freeze is shorter like around 5s, so I never noticed it before )

After adding a bunch of `printf`s it seems the hang occurs at `loopThread.join();` for main thread( moving it to after disconnecting the signal did not help neither). and for the runloop thread itself, it's waiting for `app->run()` to return.

I concur with your observation, it seems somehow inside `app->run()` the GTK runloop is "stuck". (I also observed that in my case causing a redraw such as toggling on and off my fullscreen terminal also "unstuck" IntelliJ).

the only quick solution I can think of is to `loopThread.detach();` instead of `join`. (i.e. instead of waiting for it to finish; we take solace that the quit signal would eventually be processed and the thread cleaned up). not the best solution, but since IntelliJ is quitting...

**Comment by zliuva** [comment]
I think the root cause is probably my misunderstanding of GTK threading model. currently we "dispatch" main loop to its own thread, while the creation (and attempted termination) of the `app` object happens in the main thread.

accoridng to https://docs.gtk.org/gtk4/question_index.html (no. 6), this is not supported.

I'll have to learn more about this topic and refactor the code for a long term fix.

**Comment by zliuva** [comment]
Hi @weisJ I have a fix in https://github.com/weisJ/auto-dark-mode/compare/master...zliuva:gtk-stop-freeze?expand=1 let me know what you think.

**Comment by weisJ** [comment]
Looks good to me. Though I was wondering whether we should forgo giving the application an id. My thinking is that in case two IDEA based applications with the plugin installed are started they would both receive the same Gtk::Application instance (at least that is how I understand the id system to work). Then closing one application would probably mean the other doesn't receive any notifications anymore.

**Comment by zliuva** [comment]
yeah that crossed my mind too. I'm fixin' to test that as I refactor the code to extract the mini GTK app to it's own class running in the same thread. will verify as I come across that.

one thought about that (if app is reused) is that we can skip `app->quit` and let the hold and release reference counting do its thing. (in the current form it also works without the quit line. I added it as an extra precaution)

**Comment by weisJ** [comment]
> one thought about that (if app is reused) is that we can skip `app->quit` and let the hold and release reference counting do its thing. (in the current form it also works without the quit line. I added it as an extra precaution)

That is quite interesting. Not sure though whether it buys us anything (depends on how "heavy" a Gtk::Application is).  In the case we reuse the application how do we make sure not to double call `run` on it? Or is this irrelevant?
